### PR TITLE
Feature 3.4:  Have arangodump and arangorestore default to gzip formatted files

### PR DESCRIPTION
**Pull request by matthewvon**
This branch is in response to a customer issue where there was insufficient space on the server for a full dump.  They asked about compressed output.

The initial check-in needs additional work in two areas:

1.  arangodump has no command line option to disable use of gzip.  

2.  arangorestore is not yet properly identifying and using .gz files.  (Just needs a good gdb review.)

### CODE REVIEW

**Comment by matthewvon** [comment]
@jsteemann ... I would appreciate your early comments on ManagedDirectory.{h, cpp}.  I do not have DumpFeature.cpp and RestoreFeature.cpp ready for final review.  Thank you.

**Comment by jsteemann** [comment]
I am ok with the changes in general, but we can't make gzip the default in 3.4, because users may rely on the dump/restore handling plain .json files. The files may produced at least by arangodump may be used for other purposes except dumps as well, so producing .gz files without an alternative is not downwards-compatible.
We should introduce an option for arangodump to make it produce .gz files, and in 3.4 it should default to `false`. In 3.5 it can default to `true`, but the change must then be mentioned in the CHANGELOG and `UpgradingChanges35.md`, because it will impair downwards-compatibility.

**Comment by matthewvon** [comment]
Today's check-ins are just bug fixes.  Represents the arangorestore code delivered to customer today.  I have NOT implemented jsteeman's requested adjustments yet.  Will address those before final review.

**Comment by matthewvon** [comment]
@jsteemann Not urgent ... would you review the recent 3 commits to see if you agree they meet the expectations for 3.4.  @Simran-B helped set the setIntroducedIn() parameters.  There was an issue with Examples/arangodump.json that will likely be cleaned up in release processing.

**Comment by Simran-B** [comment]
@matthewvon does this look good to you, the example in particular? https://github.com/arangodb/arangodb/pull/8811/commits/fcf165993b55a34d5fb71309c8f37ef6098f2ece

v3.3.23 is already out, v3.4.6 is already in QA I think? I guess the version numbers need to be adjusted then?

File: arangosh/Dump/DumpFeature.cpp [hunk 1]
```diff
@@ -591,7 +591,15 @@ void DumpFeature::collectOptions(std::shared_ptr<options::ProgramOptions> option
...
   options->addOption("--maskings", "file with maskings definition",
                      new StringParameter(&_options.maskingsFile))
-                     .setIntroducedIn(30322).setIntroducedIn(30402);
+                     .setIntroducedIn(30322)
+                     .setIntroducedIn(30402);
+
+  options->addOption("--compress-output",
+                     "compress files containing collection contents using gzip format",
+                     new BooleanParameter(&_options.useGzip))
+                     .setIntroducedIn(30323)
```
Comment (jsteemann) [inline arangosh/Dump/DumpFeature.cpp]: Let's not backport this to 3.3 anymore. I think backporting to 3.4 is already enough.
```suggestion
```

File: Documentation/Books/Manual/Programs/Arangodump/Examples.md [hunk 2]
```diff
@@ -180,3 +180,29 @@ Using a different key will lead to the backup being non-recoverable.
 Note that encrypted backups can be used together with the already existing 
 RocksDB encryption-at-rest feature, but they can also be used for the MMFiles
 engine, which does not have encryption-at-rest.
+
+Compression
+-----------
+
+<small>Introduced in: v3.3.23, v3.4.6, v3.5.0</small>
```
Comment (jsteemann) [inline Documentation/Books/Manual/Programs/Arangodump/Examples.md]: Let's not backport this to 3.3 anymore please.
```suggestion
<small>Introduced in: v3.4.6, v3.5.0</small>
```

File: arangosh/Utils/ManagedDirectory.cpp [hunk 3]
```diff
@@ -387,27 +399,51 @@ VPackBuilder ManagedDirectory::vpackFromJsonFile(std::string const& filename) {
...
+    } // else
+    _gzFile = gzdopen(_gzfd, gzFlags);
+  } // if
 }
 
 ManagedDirectory::File::~File() {
   try {
+    if (_gzfd >=0) {
+      gzclose(_gzFile);
+      _gzfd = -1;
```
Comment (jsteemann) [inline arangosh/Utils/ManagedDirectory.cpp]: no need for this, as we are in the destructor already.

**Comment by matthewvon** [comment]
@jsteemann I think we are ready to merge.

**Comment by matthewvon** [comment]
@jsteemann Build is not perfect.  The errors appear unrelated to changes:  

http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/4246/

**Pull request merged**